File: ics-runtime/ics_runtime/contracts/output_contract.py
"""OutputContract — declares the expected shape of LLM output."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any, Literal, TYPE_CHECKING

if TYPE_CHECKING:
    from pydantic import BaseModel

from ics_runtime.contracts.violation import ContractViolation


@dataclass
class ValidationOutcome:
    """Result of OutputContract.validate()."""
    passed: bool
    parsed: Any = None                         # Pydantic model instance on success
    violations: list[ContractViolation] = None  # type: ignore[assignment]
    is_structured_failure: bool = False        # True if response matches a failure_mode

    def __post_init__(self) -> None:
        if self.violations is None:
            self.violations = []


class OutputContract:
    """Bridges the ICS OUTPUT_CONTRACT layer with runtime schema enforcement.

    The ``schema`` is validated post-execution by attempting to parse the
    LLM response text.  ``failure_modes`` are declared expected non-error
    states (e.g. ``"insufficient_data"``) that should NOT be treated as
    violations.

    Args:
        schema:        A Pydantic v2 ``BaseModel`` subclass.  The runtime
                       calls ``schema.model_validate_json(response_text)``
                       after extracting any JSON code-fence block.
        failure_modes: List of string prefixes/labels (e.g. ``["BLOCKED:",
                       "insufficient_data"]``).  A response starting with one
                       of these is treated as a structured failure, not a
                       violation.
        format_hint:   ``"json"`` (default) or ``"markdown"``.  Determines how
                       the response is parsed.
        strict:        If ``True``, raise ``ContractViolationError`` directly
                       from ``validate()``.  Defaults to ``False`` so that the
                       caller can decide.
    """

    def __init__(
        self,
        schema: type | None = None,
        failure_modes: list[str] | None = None,
        format_hint: Literal["json", "markdown"] = "json",
        strict: bool = False,
    ) -> None:
        self.schema = schema
        self.failure_modes = failure_modes or []
        self.format_hint = format_hint
        self.strict = strict

    # ------------------------------------------------------------------
    # ICS layer text generation
    # ------------------------------------------------------------------

    def to_ics_text(self) -> str:
        """Render this contract as OUTPUT_CONTRACT ICS layer content."""
        lines: list[str] = []
        if self.schema:
            try:
                schema_json = self.schema.model_json_schema()
                lines.append(f"FORMAT: json")
                lines.append(f"SCHEMA:\n{json.dumps(schema_json, indent=2)}")
            except Exception:
                lines.append(f"FORMAT: {self.format_hint}")
        else:
            lines.append(f"FORMAT: {self.format_hint}")
        if self.failure_modes:
            lines.append("FAILURE_MODES: " + ", ".join(self.failure_modes))
        return "\n".join(lines)

    # ------------------------------------------------------------------
    # Validation
    # ------------------------------------------------------------------

    def validate(self, response_text: str) -> ValidationOutcome:
        """Validate ``response_text`` against this contract.

        Never raises — violations are recorded in the returned
        ``ValidationOutcome``.  Use ``RunResult.raise_on_violation()``
        if you want exceptions.
        """
        text = response_text.strip()

        # Check failure modes first
        for fm in self.failure_modes:
            if text.startswith(fm):
                return ValidationOutcome(passed=True, is_structured_failure=True)

        if self.schema is None:
            return ValidationOutcome(passed=True)

        if self.format_hint == "markdown":
            return ValidationOutcome(passed=True)

        # Attempt JSON extraction and schema validation
        try:
            json_text = _extract_json(text)
            parsed_dict = json.loads(json_text)
            instance = self.schema.model_validate(parsed_dict)
            return ValidationOutcome(passed=True, parsed=instance)
        except json.JSONDecodeError as exc:
            v = ContractViolation(
                rule=f"OUTPUT_CONTRACT: response is not valid JSON",
                kind="schema",
                severity="detected",
                evidence=text[:200],
            )
            return ValidationOutcome(passed=False, violations=[v])
        except Exception as exc:
            # Pydantic ValidationError or other parse failure
            violations = _pydantic_violations(exc, text)
            return ValidationOutcome(passed=False, violations=violations)


# ------------------------------------------------------------------
# Helpers
# ------------------------------------------------------------------

def _extract_json(text: str) -> str:
    """Extract JSON from a plain string or a markdown ```json ... ``` fence."""
    # Try to strip ```json ... ``` fences
    m = re.search(r"```(?:json)?\s*(\{.*?\}|\[.*?\])\s*```", text, re.DOTALL)
    if m:
        return m.group(1)
    # Fall through: assume the whole text is JSON
    return text


def _pydantic_violations(exc: Exception, response_text: str) -> list[ContractViolation]:
    violations: list[ContractViolation] = []
    try:
        for error in exc.errors():  # type: ignore[attr-defined]
            field_path = ".".join(str(loc) for loc in error.get("loc", []))
            violations.append(ContractViolation(
                rule=f"OUTPUT_CONTRACT schema: {error.get('msg', str(error))}",
                kind="schema",
                severity="detected",
                evidence=response_text[:120],
                field=field_path,
            ))
    except Exception:
        violations.append(ContractViolation(
            rule=f"OUTPUT_CONTRACT schema: {exc}",
            kind="schema",
            severity="detected",
            evidence=response_text[:120],
        ))
    return violations

File: ics-runtime/java/src/main/java/io/ics/runtime/contracts/OutputContract.java
package io.ics.runtime.contracts;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.ics.runtime.Violation;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Bridges the ICS OUTPUT_CONTRACT layer with runtime schema enforcement.
 *
 * <p>In Java, since we don't have Pydantic, JSON Schema validation is done by
 * checking required fields and basic types against a user-supplied schema map.
 * You can also register a custom {@link ResponseValidator} for full validation.
 *
 * <pre>{@code
 * OutputContract contract = OutputContract.builder()
 *     .requiredFields("decision", "score", "risk_category", "lead_id", "rationale")
 *     .failureMode("BLOCKED:")
 *     .failureMode("insufficient_data")
 *     .validator(json -> {
 *         String decision = (String) json.get("decision");
 *         if (!Set.of("QUALIFIED","NOT_QUALIFIED","REVIEW_REQUIRED").contains(decision)) {
 *             return List.of(new Violation("OUTPUT_CONTRACT: invalid decision value", ...));
 *         }
 *         return List.of();
 *     })
 *     .build();
 * }</pre>
 */
public final class OutputContract {

    @FunctionalInterface
    public interface ResponseValidator {
        /**
         * Validate the parsed JSON object.
         * @return List of violations (empty = pass)
         */
        List<Violation> validate(Map<String, Object> parsed);
    }

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final Pattern JSON_FENCE = Pattern.compile(
            "```(?:json)?\\s*(\\{.*\\}|\\[.*\\])\\s*```", Pattern.DOTALL);

    private final List<String> requiredFields;
    private final List<String> failureModes;
    private final ResponseValidator validator;
    private final String formatHint;

    private OutputContract(Builder b) {
        this.requiredFields = List.copyOf(b.requiredFields);
        this.failureModes   = List.copyOf(b.failureModes);
        this.validator      = b.validator;
        this.formatHint     = b.formatHint;
    }

    // ── ICS layer text ──────────────────────────────────────────────────────

    /**
     * Render this contract as OUTPUT_CONTRACT ICS layer content for the prompt.
     */
    public String toIcsText() {
        StringBuilder sb = new StringBuilder();
        sb.append("FORMAT: ").append(formatHint).append("\n");
        if (!requiredFields.isEmpty()) {
            sb.append("REQUIRED_FIELDS: ").append(String.join(", ", requiredFields)).append("\n");
        }
        if (!failureModes.isEmpty()) {
            sb.append("FAILURE_MODES: ").append(String.join(", ", failureModes)).append("\n");
        }
        return sb.toString().trim();
    }

    // ── Validation ──────────────────────────────────────────────────────────

    /**
     * Validate {@code responseText} against this contract.
     * Never throws — violations are recorded in the returned {@link ValidationOutcome}.
     */
    public ValidationOutcome validate(String responseText) {
        String text = responseText.strip();

        // Check failure modes first
        for (String fm : failureModes) {
            if (text.startsWith(fm)) {
                return ValidationOutcome.structuredFailure();
            }
        }

        if (!"json".equals(formatHint)) {
            return ValidationOutcome.ok(null);
        }

        // Extract JSON
        Map<String, Object> parsed;
        try {
            String jsonText = extractJson(text);
            parsed = JSON.readValue(jsonText, new TypeReference<>(){});
        } catch (Exception e) {
            return ValidationOutcome.failed(List.of(new Violation(
                    "OUTPUT_CONTRACT: response is not valid JSON",
                    "schema", "detected", text.substring(0, Math.min(200, text.length())))));
        }

        // Check required fields
        List<Violation> violations = new ArrayList<>();
        for (String field : requiredFields) {
            if (!parsed.containsKey(field)) {
                violations.add(new Violation(
                        "OUTPUT_CONTRACT: missing required field '" + field + "'",
                        "schema", "detected",
                        text.substring(0, Math.min(120, text.length())),
                        field));
            }
        }

        // Custom validator
        if (violations.isEmpty() && validator != null) {
            violations.addAll(validator.validate(parsed));
        }

        if (!violations.isEmpty()) {
            return ValidationOutcome.failed(violations);
        }
        return ValidationOutcome.ok(parsed);
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    private static String extractJson(String text) {
        Matcher m = JSON_FENCE.matcher(text);
        if (m.find()) return m.group(1);
        return text;
    }

    // ── Builder ─────────────────────────────────────────────────────────────

    public static Builder builder() { return new Builder(); }

    public static final class Builder {
        private final List<String> requiredFields = new ArrayList<>();
        private final List<String> failureModes   = new ArrayList<>();
        private ResponseValidator validator = null;
        private String formatHint = "json";

        public Builder requiredFields(String... fields) {
            requiredFields.addAll(Arrays.asList(fields));
            return this;
        }
        public Builder failureMode(String prefix) {
            failureModes.add(prefix); return this;
        }
        public Builder validator(ResponseValidator v) {
            this.validator = v; return this;
        }
        public Builder formatHint(String v) {
            this.formatHint = v; return this;
        }
        public OutputContract build() { return new OutputContract(this); }
    }
}
